plugin/pkg/shoot/quotavalidator/admission.go

```diff
@@ -83,15 +123,454 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
+	if shoot.DeletionTimestamp != nil && shoot.Annotations[common.ConfirmationDeletionTimestamp] != "" {
+		return nil
+	}
+
+	quotaReferences, err := h.getShootsQuotaReferences(*shoot)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	// Quotas are cumulative, means each quota must be not exceeded that the admission pass.
+	var maxShootLifetime *int
+	for _, quotaRef := range quotaReferences {
+		quota, err := h.quotaLister.
+			Quotas(quotaRef.Namespace).
+			Get(quotaRef.Name)
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+
+		// Get the max clusterLifeTime
+		if quota.Spec.ClusterLifetimeDays != nil {
+			if maxShootLifetime == nil {
+				maxShootLifetime = quota.Spec.ClusterLifetimeDays
+			}
+			if *maxShootLifetime > *quota.Spec.ClusterLifetimeDays {
+				maxShootLifetime = quota.Spec.ClusterLifetimeDays
+			}
+		}
+
+		exceededMetrics, err := h.isQuotaExceeded(*shoot, *quota)
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+		if exceededMetrics != nil {
+			var message bytes.Buffer
+			for _, metric := range *exceededMetrics {
+				message.WriteString(metric.String() + " ")
+			}
+			return admission.NewForbidden(a, fmt.Errorf("Quota limits exceeded. Unable to allocate further %s", message.String()))
+		}
+	}
+
+	if lifetime, exists := shoot.Annotations[common.ClusterExpirationTime]; exists && maxShootLifetime != nil {
+		var (
+			plannedExpirationTime   time.Time
+			oldExpirationTime       time.Time
+			calulatedExpirationTime time.Time
+		)
+
+		plannedExpirationTime, err = time.Parse(time.RFC3339, lifetime)
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+
+		// Get the prior version of the Shoot
+		oldShoot, ok := a.GetOldObject().(*garden.Shoot)
+		if !ok {
+			return apierrors.NewBadRequest("could not convert resource into Shoot object")
+		}
+
+		if lifetime, exists := oldShoot.Annotations[common.ClusterExpirationTime]; exists {
+			oldExpirationTime, err = time.Parse(time.RFC3339, lifetime)
+			if err != nil {
+				return apierrors.NewInternalError(err)
+			}
+		} else {
+			oldExpirationTime = shoot.CreationTimestamp.Time
+		}
+
+		calulatedExpirationTime = oldExpirationTime.Add(time.Duration(*maxShootLifetime*24) * time.Hour)
+		if plannedExpirationTime.After(calulatedExpirationTime) {
+			return admission.NewForbidden(a, fmt.Errorf("Requested shoot expiration time to long. Can only be extended by %d day(s)", *maxShootLifetime))
+		}
 	}
 
 	return nil
 }
+
+func (h *RejectShootIfQuotaExceeded) getShootsQuotaReferences(shoot garden.Shoot) ([]v1.ObjectReference, error) {
+	switch shoot.Spec.Cloud.SecretBindingRef.Kind {
+	case "CrossSecretBinding":
+		usedCrossSB, err := h.crossSBLister.
+			CrossSecretBindings(shoot.Namespace).
+			Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+		if err != nil {
+			return nil, err
+		}
+		return usedCrossSB.Quotas, nil
+	case "PrivateSecretBinding":
+		usedPrivateSB, err := h.privateSBLister.
```

Comment:
Can we write these statements into one line (line breakings are not used anywhere else in the code)?

Reply:
done